staging/src/k8s.io/apiserver/pkg/storage/cacher/interval_tracker.go

```diff
@@ -0,0 +1,99 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cacher
+
+import "container/heap"
+
+// intervalTracker keeps track of watchCacheIntervals based on
+// the startIndex of each of these intervals. A min-heap sorted
+// based on startIndex is constructed from the intervals being
+// tracked.
+//
+// The main purpose of tracking intervals in this manner is to
+// be able to find all intervals that consist of an index in its
+// range [startIndex, endIndex) and invalidate them.
+type intervalTracker struct {
+	intervals []*watchCacheInterval
+}
+
+func newIntervalTracker() *intervalTracker {
+	return &intervalTracker{
+		intervals: []*watchCacheInterval{},
+	}
+}
+
+func (it intervalTracker) Len() int { return len(it.intervals) }
+func (it intervalTracker) Less(i, j int) bool {
+	return it.intervals[i].startIndex < it.intervals[j].startIndex
+}
+func (it intervalTracker) Swap(i, j int) {
+	it.intervals[i], it.intervals[j] = it.intervals[j], it.intervals[i]
+}
+
+func (it *intervalTracker) Push(x interface{}) {
+	interval := x.(*watchCacheInterval)
+	it.intervals = append(it.intervals, interval)
+}
+
+func (it *intervalTracker) Pop() interface{} {
+	old := it.intervals
+	n := len(old)
+	interval := old[n-1]
+	old[n-1] = nil
+	it.intervals = old[0 : n-1]
+	return interval
+}
+
+var _ heap.Interface = (*intervalTracker)(nil)
```

Comment:
Tracking doesn't work as of now, because as part of regular operations on the interval (filling the buffer), you actually modify the start index (which then invalidates the invariants of the heap).

Thinking more about that - I think that we went with over-complicated path.
Let me throw a different idea to approach this.

- in watchCacheInterval.Next(), you're locking on watchCache.Lock anyway to check if the interval is valid
- so basically, what we can do is to simply not have any interval tracker here, but instead:
  (a) add isIndexValid(index) method to watchCache
  (b) when filling the buffer of watchCacheInterval, simply check if its initial index is still valid

To should reduce the amount of code by 100+ lines of code

Reply:
That's a great catch, thanks @wojtek-t!
I've tried to implement this in the latest commit https://github.com/kubernetes/kubernetes/pull/105483/commits/91dcbabbf3556743589ecc1c584cb6bf54c71f0e, PTAL whenever you can and please lmk if that looks good. If things on a high level look good, I can mark the PR ready for review and test against the CI and subsequently start adding test coverage as well :)